srml/contracts/src/lib.rs

```diff
@@ -431,11 +431,13 @@ decl_module! {
 			// can alter the balance of the caller.
 			gas::refund_unused_gas::<T>(&origin, gas_meter, imbalance);
 
-			// Dispatch every recorded call with an appropriate origin.
-			ctx.calls.into_iter().for_each(|(who, call)| {
-				let result = call.dispatch(RawOrigin::Signed(who.clone()).into());
-				Self::deposit_event(RawEvent::Dispatched(who, result.is_ok()));
-			});
+			if let Ok(_) = result {
+				// Dispatch every recorded call with an appropriate origin.
+				ctx.calls.into_iter().for_each(|(who, call)| {
+					let result = call.dispatch(RawOrigin::Signed(who.clone()).into());
+					Self::deposit_event(RawEvent::Dispatched(who, result.is_ok()));
+				});
+			}
```

Comment:
I'm not sure that helps, I mean this calls and events are always empty at this stage ? 
If those are for obvious reason empty it is better not to process them and write a comment or an assert with the proof.
I should read the code again to see why.

Reply:
So you mean an assert like: 
```
assert!(false == result.is_err() && !ctx.calls.is_empty(), 
    "Contract execution failed, no calls are dispatched")
```

Reply:
I was confused and thoughts that calls were empty on success as well. but no indeed.

I finally understand the code :-), if a failing call is guarantee to not change overlay, events and calls, then this could be done:
```diff
iff --git a/srml/contracts/src/lib.rs b/srml/contracts/src/lib.rs
index 58575ef89..e7cd79f37 100644
--- a/srml/contracts/src/lib.rs
+++ b/srml/contracts/src/lib.rs
@@ -412,18 +412,16 @@ decl_module! {
 
 			let result = ctx.call(dest, value, &mut gas_meter, &data, exec::EmptyOutputBuf::new());
 
-			if let Ok(_) = result {
-				// Commit all changes that made it thus far into the persistent storage.
-				DirectAccountDb.commit(ctx.overlay.into_change_set());
-
-				// Then deposit all events produced.
-				ctx.events.into_iter().for_each(|indexed_event| {
-					<system::Module<T>>::deposit_event_indexed(
-						&*indexed_event.topics,
-						<T as Trait>::Event::from(indexed_event.event).into(),
-					);
-				});
-			}
+			// Commit all changes that made it thus far into the persistent storage.
+			DirectAccountDb.commit(ctx.overlay.into_change_set());
+
+			// Then deposit all events produced.
+			ctx.events.into_iter().for_each(|indexed_event| {
+				<system::Module<T>>::deposit_event_indexed(
+					&*indexed_event.topics,
+					<T as Trait>::Event::from(indexed_event.event).into(),
+				);
+			});
 
 			// Refund cost of the unused gas.
 			//

+ same in other method
```

but I'm ok with this PR as it is now, it is a double check but it probably doesn't harm.

Reply:
oh I saw the revert but I might havn't express my precisely I meant either we get rid of all `if let Ok(_)` as all context is empty (overlay, events and calls) either we double check all of them. But I don't know what is prefered

Reply:
I am not sure if that worth changing now. We are about to make a refactor of the executive module anyway.

Reply:
So are we happy with this PR just to be the test to prove that this is not a bug?

Reply:
The test is the primary thing, I'd say, yes.

Reply:
Sounds like an approval to me :)